x-pack/platform/packages/shared/onechat/onechat-genai-utils/tools/search/graph.ts

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { StateGraph, Annotation } from '@langchain/langgraph';
+import type { BaseMessage } from '@langchain/core/messages';
+import { isToolMessage } from '@langchain/core/messages';
+import { messagesStateReducer } from '@langchain/langgraph';
+import { ToolNode } from '@langchain/langgraph/prebuilt';
+import type { ScopedModel } from '@kbn/onechat-server';
+import type { ElasticsearchClient } from '@kbn/core/server';
+import type { ToolResult } from '@kbn/onechat-common/tools';
+import { createNaturalLanguageSearchTool, createRelevanceSearchTool } from './inner_tools';
+import { getSearchPrompt } from './prompts';
+
+const StateAnnotation = Annotation.Root({
+  // inputs
+  nlQuery: Annotation<string>(),
+  index: Annotation<string>(),
+  // inner
+  indexIsValid: Annotation<boolean>(),
+  messages: Annotation<BaseMessage[]>({
+    reducer: messagesStateReducer,
+    default: () => [],
+  }),
+  // outputs
+  error: Annotation<string>(),
+  results: Annotation<ToolResult[]>({
+    reducer: (a, b) => [...a, ...b],
+    default: () => [],
+  }),
+});
+
+export type StateType = typeof StateAnnotation.State;
+
+export const createSearchToolGraph = ({
+  model,
+  esClient,
+}: {
+  model: ScopedModel;
+  esClient: ElasticsearchClient;
+}) => {
+  const tools = [
+    createRelevanceSearchTool({ model, esClient }),
+    createNaturalLanguageSearchTool({ model, esClient }),
+  ];
+
+  const toolNode = new ToolNode<typeof StateAnnotation.State.messages>(tools);
+
+  const checkIndex = async (state: StateType) => {
+    let index = state.index;
+    if (!index.startsWith('*')) {
+      index = `*${index}`;
+    }
+    if (!index.endsWith('*')) {
+      index = `${index}*`;
+    }
```

Comment:
AFAICT, this isn't persisted, so this can lead to us deciding an index is valid (because `*test*` matches something), but then when we actually try to search without the leading and trailing `*` it'll fail, right?

Reply:
Yeah that's correct. 

I left a comment on that same code saying it's a very naive implem. I can switch to "hard checking" if an index exists instead if we prefer. In practice there is no real risks atm because the tool will (well, should) be called with a single, valid, index name, so I assumed it was fine for now.

One of the next big "steps" for onechat is index sources / restraining agents on index patterns, and I suspect that code will highly evolve then (which is also why I didn't want to spend too much time on it).

Reply:
EDIT: fixed it as part of https://github.com/elastic/kibana/pull/232356/commits/598cf8b40b35c78f46bd5a9ee33e21adc4d849f6